nimblehq/rails-templates
# Upgrade to Ruby 3.0.1

https://github.com/nimblehq/rails-templates/issues/309

## What happened

- Upgrade Ruby from `2.7.2` to `3.0.1`
- Upgrade rspec-rails to `5.0.2`
- Pin parser version to match our ruby version to fix the [parser warnings](https://github.com/nimblehq/rails-templates/pull/307#issuecomment-1009567503)
- Add `.tool-versions` file to support [asdf](https://github.com/asdf-vm/asdf)

## Insight

I also tried upgrading to Ruby 3.1.0 but faced the issue:

> /.asdf/installs/ruby/3.1.0/lib/ruby/3.1.0/psych/visitors/to_ruby.rb:430:in `visit_Psych_Nodes_Alias': Unknown alias: default (Psych::BadAlias)

After investigation, I found it is from [figaro](https://github.com/laserlemon/figaro) gem.

```
from /.asdf/installs/ruby/3.1.0/lib/ruby/gems/3.1.0/gems/figaro-1.2.0/lib/figaro/application.rb:60:in `parse'
from /.asdf/installs/ruby/3.1.0/lib/ruby/gems/3.1.0/gems/figaro-1.2.0/lib/figaro/application.rb:56:in `block in raw_configuration'
from /.asdf/installs/ruby/3.1.0/lib/ruby/gems/3.1.0/gems/figaro-1.2.0/lib/figaro/application.rb:56:in `raw_configuration'
from /.asdf/installs/ruby/3.1.0/lib/ruby/gems/3.1.0/gems/figaro-1.2.0/lib/figaro/application.rb:64:in `global_configuration'
from /.asdf/installs/ruby/3.1.0/lib/ruby/gems/3.1.0/gems/figaro-1.2.0/lib/figaro/application.rb:32:in `configuration'
from /.asdf/installs/ruby/3.1.0/lib/ruby/gems/3.1.0/gems/figaro-1.2.0/lib/figaro/application.rb:42:in `each'
from /.asdf/installs/ruby/3.1.0/lib/ruby/gems/3.1.0/gems/figaro-1.2.0/lib/figaro/application.rb:36:in `load'
```

This is because on Ruby 3.1.0, it comes with Psych 4 - https://www.ruby-lang.org/en/news/2021/12/25/ruby-3-1-0-released/

> Psych 4.0 changes Psych.load to use safe_load by default. You may need to use Psych 3.3.2 for migrating to this behavior. [[Bug #17866]](https://bugs.ruby-lang.org/issues/17866)

In the issue, they explained that now YML is loaded with `aliases: false` by default. So when the app tries to load Figaro's `config/application.yml`, it failed because in that file, we're using alias. To fix we need to pass `aliases: true` when [loading the config](https://github.com/laserlemon/figaro/blob/6ae998e8209a08f91a165822abecffc0a0626814/lib/figaro/application.rb#L60) just like they've done on [Rails](https://github.com/rails/rails/commit/179d0a1f474ada02e0030ac3bd062fc653765dbe). 

So in order to upgrade to Ruby 3.1, we need to have this issue fixed on figaro. I'll open the issue. But if no one responds, we will need to fork or look into other alternatives. 

## Proof Of Work

To test, 

1. Install Ruby 3.0.1
2. Install Rails 7.0.1
3. Create a project from a template

> rails new $APP_NAME -m https://raw.githubusercontent.com/nimblehq/rails-templates/chore/upgrade-to-ruby-3/template.rb

4. Start the server and it should run perfectly

![Screen Shot 2565-01-18 at 16 12 36](https://user-images.githubusercontent.com/6965195/149906988-57157b4e-e5e8-4585-bd19-dd3e3bebdeda.png)
